mapeditor/mapview.cpp

```diff
@@ -0,0 +1,448 @@
+			int3{-1, 1, 0}, int3{0, 1, 0}, int3{1, 1, 0}, int3{2, 1, 0},
+			int3{-1, 2, 0}, int3{0, 2, 0}, int3{1, 2, 0}, int3{2, 2, 0}
+		};
+		for(auto & e : extra)
+		{
+			if(mouseEvent->buttons() & Qt::RightButton)
+				sc->selectionTerrainView.erase(tile + e);
+			else if(mouseEvent->buttons() == Qt::LeftButton)
+				sc->selectionTerrainView.select(tile + e);
+		}
+	}
+		sc->selectionTerrainView.draw();
+		break;
+
+	case MapView::SelectionTool::Area:
+		if(mouseEvent->buttons() & Qt::RightButton || !mouseEvent->buttons() & Qt::LeftButton)
+			break;
+
+		sc->selectionTerrainView.clear();
+		for(int j = std::min(tile.y, tileStart.y); j < std::max(tile.y, tileStart.y); ++j)
+		{
+			for(int i = std::min(tile.x, tileStart.x); i < std::max(tile.x, tileStart.x); ++i)
+			{
+				sc->selectionTerrainView.select(int3(i, j, sc->level));
+			}
+		}
+		sc->selectionTerrainView.draw();
+		break;
+
+	case MapView::SelectionTool::None:
+		if(mouseEvent->buttons() & Qt::RightButton)
+			break;
+
+		auto sh = tile - tileStart;
+		sc->selectionObjectsView.shift = QPoint(sh.x, sh.y);
+
+		if(sh.x || sh.y)
+		{
+			if(sc->selectionObjectsView.newObject)
+			{
+				sc->selectionObjectsView.shift = QPoint(tile.x, tile.y);
+				sc->selectionObjectsView.selectObject(sc->selectionObjectsView.newObject);
+				sc->selectionObjectsView.selectionMode = 2;
+			}
+			else if(mouseEvent->buttons() & Qt::LeftButton)
+			{
+				if(sc->selectionObjectsView.selectionMode == 1)
+				{
+					sc->selectionObjectsView.clear();
+					sc->selectionObjectsView.selectObjects(tileStart.x, tileStart.y, tile.x, tile.y);
+				}
+			}
+		}
+
+		sc->selectionObjectsView.draw();
+		break;
+	}
+}
+
+void MapView::mousePressEvent(QMouseEvent *event)
+{
+	this->update();
+
+	auto * sc = static_cast<MapScene*>(scene());
+	if(!sc || !controller->map())
+		return;
+
+	mouseStart = mapToScene(event->pos());
+	tileStart = tilePrev = int3(mouseStart.x() / 32, mouseStart.y() / 32, sc->level);
+
+	if(sc->selectionTerrainView.selection().count(tileStart))
+		pressedOnSelected = true;
+	else
+		pressedOnSelected = false;
+
+	switch(selectionTool)
+	{
+	case MapView::SelectionTool::Brush:
+		sc->selectionObjectsView.clear();
+		sc->selectionObjectsView.draw();
+
+		if(event->button() == Qt::RightButton)
+			sc->selectionTerrainView.erase(tileStart);
+		else if(event->button() == Qt::LeftButton)
+			sc->selectionTerrainView.select(tileStart);
+		sc->selectionTerrainView.draw();
+		break;
+
+	case MapView::SelectionTool::Brush2:
+		sc->selectionObjectsView.clear();
+		sc->selectionObjectsView.draw();
+	{
+		std::array<int3, 4> extra{ int3{0, 0, 0}, int3{1, 0, 0}, int3{0, 1, 0}, int3{1, 1, 0} };
+		for(auto & e : extra)
+		{
+			if(event->button() == Qt::RightButton)
+				sc->selectionTerrainView.erase(tileStart + e);
+			else if(event->button() == Qt::LeftButton)
+				sc->selectionTerrainView.select(tileStart + e);
+		}
+	}
+		sc->selectionTerrainView.draw();
+		break;
+
+	case MapView::SelectionTool::Brush4:
```

Comment:
Looks like these switch/cases differ only with brush size

Reply:
do you suggest short lambda?